File: src/com/OpenSerenity/elements/WdNativeElement.java
package com.OpenSerenity.elements;

import com.OpenSerenity.core.TestContext;
import com.OpenSerenity.functionalInterfaces.Func;
import com.OpenSerenity.functionalInterfaces.WaitCondition;
import com.OpenSerenity.utils.Waiter;
import org.openqa.selenium.By;
import org.openqa.selenium.StaleElementReferenceException;
import org.openqa.selenium.WebElement;

import java.util.ArrayList;
import java.util.List;

public class WdNativeElement implements NativeElement {
    private Func<WebElement> wdElementSelector;

    private WebElement wdElement;
    WebElement getWdElement() throws Exception {
        if(isStale()) {
            wdElement = wdElementSelector.invoke();
        }
        return wdElement;
    }

    public WdNativeElement(Func<WebElement> wdElementSelector) {
        this.wdElementSelector = wdElementSelector;
    }

    @Override
    public String getTagName() throws Exception {
        try {
            return wdElement.getTagName();
        } catch (Exception ex) {
            ex.printStackTrace();
            return "";
        }
    }

    @Override
    public String getText() throws Exception {
        return getWdElement().getText();
    }

    @Override
    public boolean isEnabled() throws Exception {
        return getWdElement().isEnabled();
    }

    @Override
    public boolean isSelected() throws Exception {
        return getWdElement().isSelected();
    }

    @Override
    public boolean isDisplayed() throws Exception {
        return getWdElement().isDisplayed();
    }

    @Override
    public boolean isExists() {
        try
        {
            return getWdElement() != null;
        }
        catch (Exception ex)
        {
            ex.printStackTrace();
            return false;
        }
    }

    @Override
    public void clear() throws Exception {
        getWdElement().clear();
    }

    @Override
    public void sendKeys(String text) throws Exception {
        getWdElement().sendKeys(text);
    }

    @Override
    public void submit() throws Exception {
        getWdElement().submit();
    }

    @Override
    public void click() throws Exception {
        getWdElement().click();
    }

    @Override
    public void altClick() {
        throw new UnsupportedOperationException(); //To change body of implemented methods use File | Settings | File Templates.
    }

    @Override
    public String getAttribute(String attributeName) throws Exception {
        try {
            return wdElement.getAttribute(attributeName);
        } catch (Exception ex) {
            ex.printStackTrace();
            return "";
        }
    }

    @Override
    public String getCssValue(String propertyName) throws Exception {
        try {
            return wdElement.getCssValue(propertyName);
        } catch (Exception ex) {
            ex.printStackTrace();
            return "";
        }
    }

    @Override
    public NativeElement findChild(final String locator) {
        return new WdNativeElement(new Func<WebElement>() {
            @Override
            public WebElement invoke() throws Exception {
                Waiter.WithTimeout(TestContext.configuration.getElementFindTimeout()).waitFor(new WaitCondition() {
                    @Override
                    public boolean invoke() throws Exception {
                        return getWdElement().findElement(By.cssSelector(locator)) != null;
                    }
                });
                return getWdElement().findElement(By.cssSelector(locator));
            }
        });
    }

    @Override
    public List<NativeElement> findChildren(final String locator) {
        List<NativeElement> result = new ArrayList<>();
        try {
            List<WebElement> wdElementsList = getWdElement().findElements(By.cssSelector(locator));

            for(final WebElement webElement : wdElementsList) {
                result.add(new WdNativeElement(new Func<WebElement>() {
                    @Override
                    public WebElement invoke() throws Exception {
                        return webElement;
                    }
                }));
            }

            return result;
        } catch (Exception e) {
            return new ArrayList<>();
        }
    }

    public boolean isStale() {
        try {
            if(wdElement == null) {
                return true;
            }
            wdElement.findElement(By.xpath("."));
            boolean res = wdElement.isDisplayed();
            return false;
        } catch (StaleElementReferenceException ex) {
            ex.printStackTrace();
            return true;
        }
    }
}

File: src/scala/com/OpenSerenity/elements/WdNativeElement.scala
package com.OpenSerenity.elements

import org.openqa.selenium.{StaleElementReferenceException, By, WebElement}
import scala.Predef.String
import com.OpenSerenity.utils.Waiter
import com.OpenSerenity.core.TestContext

class WdNativeElement extends NativeElement {
  private[elements] def getWdElement: WebElement = {
    if (isStale) {
      wdElement = wdElementSelector()
    }
    wdElement
  }

  def this(wdElementSelector: () => WebElement) {
    this()
    this.wdElementSelector = wdElementSelector
  }

  def getTagName: String = {
    try {
      return wdElement.getTagName
    }
    catch {
      case ex: Exception => {
        ex.printStackTrace
        return ""
      }
    }
  }

  def getText = getWdElement.getText.trim
  def isEnabled = getWdElement.isEnabled
  def isSelected = getWdElement.isSelected
  def isDisplayed = getWdElement.isDisplayed

  def isExists = {
    try {
      getWdElement != null
    }
    catch {
      case ex: Exception => {
        ex.printStackTrace
        false
      }
    }
  }

  def clear: Unit = getWdElement.clear()
  def sendKeys(text: String): Unit = getWdElement.sendKeys(text)
  def submit: Unit = getWdElement.submit()
  def click: Unit = getWdElement.click()
  def altClick = throw new UnsupportedOperationException

  def getAttribute(attributeName: String) = {
    try {
      getWdElement.getAttribute(attributeName)
    }
    catch {
      case ex: Exception => {
        ex.printStackTrace
        ""
      }
    }
  }

  def getCssValue(propertyName: String) = {
    try {
      getWdElement.getCssValue(propertyName)
    }
    catch {
      case ex: Exception => {
        ex.printStackTrace
        ""
      }
    }
  }

  def findChild(locator: String): NativeElement = {
    return new WdNativeElement(() => {
        Waiter
          .withTimeout(TestContext.configuration.getElementFindTimeout)
          .waitFor(() => getWdElement.findElement(By.cssSelector(locator)) != null)
        getWdElement.findElement(By.cssSelector(locator))
    })
  }

  def findChildren(locator: String): List[NativeElement] = {
    throw new NotImplementedError()
  }

  def isStale = {
    try {
      if (wdElement == null) {
        true
      }
      wdElement.findElement(By.xpath("."))
      val res: Boolean = wdElement.isDisplayed
      false
    }
    catch {
      case ex: StaleElementReferenceException => {
        ex.printStackTrace
        true
      }
    }
  }

  private var wdElementSelector: () => WebElement = null
  private var wdElement: WebElement = null
}


